beps/0003-auth-architecture-evolution/README.md

```diff
@@ -130,96 +212,91 @@ export interface HttpAuthService {
   // TODO: user can currently be undefined, figure out best pattern to avoid that
   const { user } = httpAuth.credentials(req);
+  if (!user) {
+    throw new NotAllowedError(
+      'Service requests are not allowed on this endpoint',
+    );
+  }
   console.log(
     `User ref=${user.userEntityRef} ownership=${user.ownershipEntityRefs}`,
   );
   // ...
 });
 ```
 
-**Forward the user credentials from an incoming requests to upstream plugin backend**
+#### Forward the user credentials from an incoming requests to upstream plugin backend
 
 ```ts
-router.get(
-  '/read-data',
-  // Here, the middleware applies to the current path only
-  httpAuth.middleware({ allow: ['user'] }),
-  (req, res) => {
-    const entity = await catalogClient.getEntityByRef(req.params.entityRef, {
-      credentials: httpAuth.credentials(req),
-    });
-    // ...
-  },
-);
+router.get('/read-data', (req, res) => {
+  // The catalogClient will have a reference to the (plugin scoped) HttpAuthService,
+  // which it uses to create the credential headers for the upstream request.
+  const entity = await catalogClient.getEntityByRef(req.params.entityRef, {
+    credentials: httpAuth.credentials(req),
+  });
+  // ...
+});
 ```
 
-**Allow both user and service request**
+#### Allow both user and service request
 
 ```ts
-router.get(
-  '/read-data',
-  httpAuth.middleware({ allow: ['user', 'service'] }),
-  (req, res) => {
-    const credentials = httpAuth.credentials(req);
-    if (credentials.user) {
-      res.json(
-        // Silly example just to highlight separate code paths for user and
-        // service requests
-        todoStore.listOwnedTodos({ owner: credentials.user.userEntityRef }),
-      );
-    } else {
-      res.json(todoStore.listTodos());
-    }
-  },
-);
+router.get('/read-data', (req, res) => {
+  const credentials = httpAuth.credentials(req);
+  if (credentials.user) {
+    res.json(
+      // Silly example just to highlight separate code paths for user and
+      // service requests
+      todoStore.listOwnedTodos({ owner: credentials.user.userEntityRef }),
+    );
+  } else {
+    res.json(todoStore.listTodos());
+  }
+});
 ```
 
-**Issuing a cookie and allowing user cookie auth on a separate endpoint**
+#### Issuing a cookie and allowing user cookie auth on a separate endpoint
 
 ```ts
-router.get('/cookie', httpAuth.middleware({ allow: ['user'] }), (req, res) => {
-  await httpAuth.issueUserCookie(res);
+router.get('/cookie', (req, res) => {
+  await httpAuth.issueUserCookie(res); // If this is a service call it'll throw
   res.json({ ok: true });
 });
 
+// Allowing cookie auth is a separate step where you call the configure method
+// of the httpRouter API in your plugin setup code.
+httpRouter.configure({
+  // In practice we can make this configuration a lot more capable, this is just a minimal example
+  allowCookieAuthOnPaths: ['/static'], // router.use('/static', cookieAuthMiddleware()) under the hood
```

Comment:
Not sure if this is the proper place to hold this discussion but one of the limitations we ran into with cookie based auth is the limited cookie size restriction (typically 4096 bytes, half of a valid token header [size limit](https://github.com/backstage/backstage/pull/22640/files#diff-dbbd80fd00b1db5c16a2498cbb1948001e3f65175afad026558eb718ec5d3d39R26)). We have some "seasoned users" that have amassed a ton of group memberships over decades of tenure that result in a huge ownership claim producing auth cookies that are rejected by the browser for being too large. 

The ownership claim "memory leak" is an internal issue that we need to figure out a way to prune but we currently workaround this with a middleware that essentially publicly exposes the cookie based endpoint (currently just the techdocs static stylesheet request which we decided is safe enough to unprotect) by injecting a server token so the downstream request to the endpoint succeeds. 

I'm wondering under this new design, would `allowCookieAuthOnPaths` only allow cookie auth or would our workaround still work where this just means allow cookie in addition to the default user + server. 

Also on a related note, it might also be worth bringing up for discussion the reliability of cookie based auth due to the discrepancy between auth header and cookie size restrictions

Reply:
Thank you for bringing this up! 👍  There's a risk this gets lost here as I'll link from the discussion issue and we can continue there.

👀 for @freben as well, regarding the token payload discussion. Probably makes sense to at the very least open up for the ability to switch out the token verification flow without service interface changes. We'd want to also prep that we endpoints for external services to treat tokens as opaque access tokens as well, instead of parsing values out of the JWTs

On the Q: our thinking is that cookie auth would imply user and service auth too. Same thing goes for unauthorized requests for things like rate limiting.

Still I'm a lil bit amazed though @kuangp, even at `4096 / 'group:default/some-pretty-long-group-name'.length` that leaves space for 100 groups by itself. Accounting for other fields and JWT parts you'd at minimum be able to get 50 different groups in there. Does feel like a bit extreme to be able to own things from that many different groups 😅. Tbh it would really be on us to provide better tools to limit that set though, so either way some form of solution is needed.